Tribler/Policies/BoostingManager.py

```diff
@@ -0,0 +1,676 @@
+
+
+        self.boosting_sources = {}
+        self.torrents = {}
+
+        local_settings = {}
+        local_settings['share_mode_target'] = self.settings.share_mode_target
+        self.session.lm.ltmgr.get_session().set_settings(local_settings)
+
+        if os.path.exists(self.settings.config_file):
+            self._logger.info("Config file %s", open(self.settings.config_file).read())
+            self.load_config()
+        else:
+            self._logger.warning("Initial config file missing")
+
+        self.session.add_observer(self.on_torrent_notify, NTFY_TORRENTS, [NTFY_UPDATE])
+
+        # TODO(emilon): Refactor this to use taskmanager
+        self.session.lm.threadpool.add_task(self._select_torrent, self.settings.initial_swarm_interval,
+                                            "CreditMining_select_init")
+        self.session.lm.threadpool.add_task(self.scrape_trackers,
+                                            self.settings.initial_tracker_interval, "CreditMining_scrape")
+        self.session.lm.threadpool.add_task(self.log_statistics,
+                                            self.settings.initial_logging_interval, "CreditMining_log_init")
+
+    @staticmethod
+    def get_instance(*args, **kw):
+        """
+        get single instance of Boostingmanager
+        """
+        if BoostingManager.__single is None:
+            BoostingManager(*args, **kw)
+        return BoostingManager.__single
+
+    @staticmethod
+    def del_instance():
+        """
+        resetting, then deleting single instance
+        """
+        BoostingManager.__single = None
+
+    def shutdown(self):
+        """
+        save configuration before stopping stuffs
+        """
+        self.save_config()
+        self._logger.info("Shutting down boostingmanager")
+
+        for sourcekey in self.boosting_sources.keys():
+            self.remove_source(sourcekey)
+
+        if self.session.lm.threadpool.is_pending_task_active("CreditMining_select_init"):
+            self.session.lm.threadpool.cancel_pending_task("CreditMining_select_init")
+        if self.session.lm.threadpool.is_pending_task_active("CreditMining_scrape"):
+            self.session.lm.threadpool.cancel_pending_task("CreditMining_scrape")
+        if self.session.lm.threadpool.is_pending_task_active("CreditMining_log_init"):
+            self.session.lm.threadpool.cancel_pending_task("CreditMining_log_init")
+
+        self.cancel_all_pending_tasks()
+        # for torrent in self.torrents.itervalues():
+        #     self.stop_download(torrent)
+
+    def get_source_object(self, sourcekey):
+        """
+        Get the actual object of the source string/key
+        """
+        # first, assume it's already source-friendly format
+        ret = self.boosting_sources.get(sourcekey, None)
+
+        # we can't find source, lets assume it's string now
+        if not ret:
+            try:  # is it channel?
+                int(sourcekey, 16)
```

Comment:
Internally, I always use the same type. However, to prevent a lot of if-else (and hexlify/unhexlify) in other place, this function can retrieve two types of input I've mentioned.